Feature tests ready for code review

SofiBretz opened the pull request:
Hello,

- We made 10 test methods. 
- All the tests passed. 
- Can you please made a code review.

Thanks,

Sofi and Thaís.

thsvr committed:
add file

thsvr committed:
add my_each test

SofiBretz committed:
added test method for my_each_with_index, test passed

thsvr committed:
add my_select test

SofiBretz committed:
added my_all? method test, no failures

thsvr committed:
add my_any? test

SofiBretz committed:
added my_none? method test, all tests passed

thsvr committed:
add my_count test and fix

SofiBretz committed:
added my_map test, all test passed

thsvr committed:
add my_inject test

thsvr committed:
add my_inject test and fix

SofiBretz committed:
added multiply_els test, all test passed

thsvr committed:
fix rubocop and stickler

bolah2009 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      array.my_all? { |x| x > 0 }
+      expect(tested_arr).to eql(default_arr)
+    end
+  end
```
Let's test for other use cases of `my_all?`. Commenting out part of the method still pass the test as shown below:

```ruby
  def my_all?(_par = nil)
    if block_given?
      my_each { |x| return false unless yield(x) }
      # elsif par.class == Class
      #   my_each { |x| return false unless x.class == par }
      # elsif par.class == Regexp
      #   my_each { |x| return false unless x =~ par }
      # elsif par.nil?
      #   my_each { |x| return false unless x }
      # else
      #   my_each { |x| return false unless x == par }
    end
    true
  end
```

bolah2009 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      array.my_any? { |x| x > 3 }
+      expect(tested_arr).to eql(default_arr)
+    end
+  end
```
Let's test for other use cases of `my_any?`. Commenting out part of the method still pass the test as shown below:

```ruby
  def my_any?(_par = nil)
    if block_given?
      my_each { |x| return true if yield(x) }
      # elsif par.class == Class
      #   my_each { |x| return true if x.class == par }
      # elsif par.class == Regexp
      #   my_each { |x| return true if x =~ par }
      # elsif par.nil?
      #   my_each { |x| return true if x }
      # else
      #   my_each { |x| return true if x == par }
    end
    # false
  end
```

bolah2009 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      array.my_none? { |x| x == 8 }
+      expect(tested_arr).to eql(default_arr)
+    end
+  end
```
Let's test for other use cases of `my_none?`. Commenting out part of the method still pass the test as shown below:

```ruby
  def my_none?(_par = nil)
    if block_given?
      my_each { |x| return false if yield(x) }
      # elsif par.class == Class
      #   my_each { |x| return false if x.class == par }
      # elsif par.class == Regexp
      #   my_each { |x| return false if x =~ par }
      # elsif par.nil?
      #   my_each { |x| return false if x }
      # else
      #   my_each { |x| return false if x == par }
    end
    true
  end
```

bolah2009 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      array.my_count { |x| x > 5 }
+      expect(tested_arr).to eql(default_arr)
+    end
+  end
```
Let's test for other use cases of `my_count`. Commenting out part of the method still pass the test as shown below:

```ruby
  def my_count(_items = nil)
    count = 0
    if block_given?
      my_each { |x| count += 1 if yield(x) == true }
      # elsif items.nil?
      #   my_each { count += 1 }
      # else
      #   my_each { |x| count += 1 if x == items }
    end
    count
  end
```

bolah2009 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      array.my_inject { |x, y| x + y }
+      expect(tested_arr).to eql(default_arr)
+    end
+  end
```
Let's test for other use cases of `my_inject`. Commenting out part of the method still pass the test as shown below:

```ruby
  def my_inject(*args)
    arr = to_a.dup
    if args[0].nil?
      memo = arr.shift
    elsif args[1].nil? && !block_given?
      symbol = args[0]
      memo = arr.shift
      # elsif args[1].nil? && block_given?
      #   memo = args[0]
      # else
      #   memo = args[0]
      #   symbol = args[1]
    end
    arr[0..-1].my_each do |elem|
      memo = if symbol
               memo.send(symbol, elem)
             else
               yield(memo, elem)
             end
    end
    memo
  end
end
```

bolah2009 reviewed (changes requested):
# Nice job with the tests :clap: 

Can we check the test of these methods, the test are not fully covered:
- `my_all?`
- `my_any?`
- `my_none?`
- `my_count`
- `my_inject`

Let's cover all use cases and submit another code review request. You can check the comments below for more details.

#### :recycle:  Make these changes and submit another code review request. :+1: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

thsvr committed:
fix my_select

thsvr committed:
fix my_all?

thsvr committed:
fix my_any?

SofiBretz committed:
my_none, my_count, my_map and my_inject fixed, all test passed :)

SofiBretz commented:
### Summary of changes made

- Rubocop ready, no conflicts with the base branch.

- Methods: my_select, my_all?, my_any?, my_none?, my_count, my_map, my_inject fixed.

- Can you please check for another code review.

Thank you,

Sofi and Thaís :)

mcrd25 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  describe '#my_all?' do
```
@bolah2009  previously mentioned that if you commented out part of the method like so:
```ruby
def my_all?(par = nil)
    if block_given?
      my_each { |x| return false unless yield(x) }
    # elsif par.class == Class
    #   my_each { |x| return false unless x.class == par }
    # elsif par.class == Regexp
    #   my_each { |x| return false unless x =~ par }
    # elsif par.nil?
    #   my_each { |x| return false unless x }
    # else
    #   my_each { |x| return false unless x == par }
    end
    true
  end
```
All tests still pass. Which should not be the case.
There is nothing wrong with your method, but you ought to have test for each scenario. You only have a test for the first scenario really. You should have an `it` block for each expected behavior.
Try implementing tests for testing Class, Regular Expression, if there is no argument passed and no block is given, if argument is passed other than RegExp and Class. (According to your implemented method)

This test does not actually tell me your method works. It only tells me part of your method works? Do you get what I'm saying? Or am I confusing you? 😅

mcrd25 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  describe '#my_any?' do
```
Same issue as with `my_all`

mcrd25 replied:
Address please 🙏

mcrd25 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  describe '#my_none?' do
```
Also same as with `my_all` see my comment to understand how to address.

mcrd25 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  describe '#my_count' do
```
Again, please add more tests to cover each scenario. see my comments for my all. 
I'm really sorry you guys have to add more tests. But it will make you better programmers. And I'd like you guys to know common testing practices which would improve your chances of getting a job! 💪 

🙏

mcrd25 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  describe '#my_map' do
```
If I were to comment out `return to_enum unless block_given?` line from method test would still pass. You need to consider when a block is not given. (Do this for all applicable methods)

mcrd25 commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  describe '#my_inject' do
```
Please cover all scenarios covered by your ifs in this method. 


[OPTIONAL] - As the following is not related to this testing project specifically. However, everything else concerning tests is required.
This method implementation may need some refactoring in the future though I see some issues when I test it myself. For instance, if first if is true 
```ruby
if args[0].nil?
      memo = arr.shift
```
`symbol` is not defined. But the following would execute anyway which is dependent on `symbol` variable.
I'd suggest you change that. 
As you can see this executes anyway and is dependent on `symbol` variable:
```ruby
arr[0..-1].my_each do |elem|
      memo = if symbol
               memo.send(symbol, elem)
             else
               yield(memo, elem)
             end
    end
```

mcrd25 reviewed (changes requested):
# Changes Required but Good Job! 🙂 

Hi @SofiBretz and @thsvr 👋,

I am not kidding when I say this, you guys did a great job! You implemented tests for every method: which is to be applauded as you only had to implement for any 6 you chose.

There are a couple of required things to address before you can get the thumbs up.

### Required Changes 
- Address review comments NOT marked by [OPTIONAL] tag
- Make your tests, in general, have negative and positive scenarios. For instance, you have tests for `my_each` when a block is given. What about when a block is not given? Please apply this to all your methods 🙏  (This sort of ties into review comments. But you definitely have to add tests for when a block is not given for `my_each` and `my_each_with_index`

Your README is good, your stickler configuration is good and all your tests currently pass! You only need to implement tests that cover all scenarios in method and you should get the 👍. I really am glad you decided to do tests for all. Shows a good work ethic. 👏 

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

SofiBretz committed:
added false and true test to my_all?

thsvr committed:
robocop add

thsvr committed:
add more tests to my_any?, my_all?,rubocop, fix some errors

SofiBretz committed:
testing the negative scenarios

thsvr committed:
add false cases and fix some errors

thsvr committed:
add false cases and fix some errors

thsvr committed:
add test to check if no block given to my_each and my_each_with_index

SofiBretz committed:
added more tests to my_count and my_map

thsvr committed:
add test to my_inject

SofiBretz committed:
checking the last details of the review, all test passed

SofiBretz commented:
### Summary of changes made

- Rubocop, all checks have passed.

- my_each, my_each with index, my_all?, my_any?, my_none?, my_count, my_map, my_inject ready, we fix all the recommendations of Bola and Maya.

- Added more tests.

- Added negative scenarios required by Maya 👍 

Can you please check for a code review?

Thanks! :)

Sofi and Thaís.

samgaco commented on spec/enumerable_methods_spec.rb:
```diff
@@ -0,0 +1,186 @@
+  end
+
+  describe '#my_all?' do
+    it 'returns true if all elements are true (or empty array)' do
+      default = array.all? { |x| x > 0 }
+      tested = array.my_all? { |x| x > 0 }
+      expect(tested).to eql(default)
```
:100:

samgaco reviewed (approved):
HI @SofiBretz and @mcrd25 :high_brightness: 

The project is approved, you may merge and submit the project completion form.

Regards.

-----------------------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:
